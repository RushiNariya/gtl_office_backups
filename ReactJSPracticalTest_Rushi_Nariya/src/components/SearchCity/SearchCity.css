@import 'https://maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css';

form {
  display: inline-block;
  margin: 1rem auto 0;
}

.pseudo-search {
  display: inline;
  position: relative;
  border: 2px solid #fff;
  border-radius: 5px;
  padding: 10px 15px;
  transition: 0.5 ease-in-out;
  outline: none;
}

input {
  border: 0;
  color: #fff;
  background-color: transparent;
  width: 400px;
}

input::placeholder {
  color: #fff;
  opacity: 1;
}

@media (max-width: 480px) {
  input {
    width: 240px;
  }
}

input:focus {
  outline: none;
}

button {
  color: white;
  border: none;
  background: none;
  cursor: pointer;
}

select {
  border: none;
}

.no-suggestions {
  color: #fff;
  padding: 0.5rem;
}

.suggestions {
  position: relative;
  border: 1px solid #999;
  color: #fff;
  font-size: 1.2rem;
  border-top-width: 0;
  list-style: none;
  margin-top: 5px;
  max-height: 143px;
  overflow-x: hidden;
  overflow-y: auto;
  padding-left: 0;
  width: inherit;
}

.suggestions::-webkit-scrollbar {
  display: none;
}
.suggestions li {
  padding: 0.5rem;
}

.suggestion-active,
.suggestions li:hover {
  background: rgba(255, 255, 255, 0.1);
  color: #fff;
  cursor: pointer;
  font-weight: 700;
}

.suggestions li:not(:last-of-type) {
  border-bottom: 1px solid #999;
}
