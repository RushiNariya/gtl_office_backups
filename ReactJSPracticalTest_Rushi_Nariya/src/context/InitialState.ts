/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-empty-function */

import { createContext } from "react";
import cityInfo from "../intefaces/cityInfo";
import weatherInfo from "../intefaces/weatherInfo";

export type initState = {
  data: weatherInfo;
  city: Array<cityInfo>;
  error: string;
  searchByCityHandler: (data: weatherInfo) => void;
  searchCityHandler: (data: cityInfo) => void;
  setWeatherDataHandler: (data: weatherInfo) => void;
  errorHandler: (error: string) => void;
  resetSetErrorHandler: () => void;
  removeCityHandler: (city: string) => void;
};
  
const initialState: initState = {
  data: {},
  city: [],
  error: '',
  searchByCityHandler: (data) => {},
  searchCityHandler: (data) => {},
  setWeatherDataHandler: (data) => {},
  errorHandler: (error) => {},
  resetSetErrorHandler: () => {},
  removeCityHandler: (city) => {},
};

export const GlobalContext = createContext(initialState);

export default initialState;