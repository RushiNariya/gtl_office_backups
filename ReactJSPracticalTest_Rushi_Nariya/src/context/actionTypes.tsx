export const SEARCH_BY_CITY = 'SEARCH_BY_CITY';
export const SET_WEATHER_DATA = 'SET_WEATHER_DATA';
export const SEARCH_CITY = 'SEARCH_CITY';
export const REMOVE_CITY = 'REMOVE_CITY';
export const SET_ERROR = 'SET_ERROR';
export const RESET_ERROR = 'RESET_ERROR';
export const REFRESH_STATE = 'REFRESH_STATE';