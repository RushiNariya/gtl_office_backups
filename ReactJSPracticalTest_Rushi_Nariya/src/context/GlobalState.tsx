/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable @typescript-eslint/no-empty-function */
import React, { useReducer, useEffect } from 'react';
import PropTypes from 'prop-types';
import AppReducer from './reducer';
import {
  searchByCityAction,
  searchCityAction,
  setWeatherDataAction,
  setErrorAction,
  resetErrorAction,
  removeCityAction,
  refreshStateAction,
} from './actions';
import weatherInfo from '../intefaces/weatherInfo';
import cityInfo from '../intefaces/cityInfo';
import { GlobalContext, initState } from './InitialState';


type props = {
  children: any;
  initialState: initState;
};

export const GlobalProvider = ({ children, initialState }: props): JSX.Element => {
  const [state, dispatch] = useReducer(AppReducer, initialState);

  function searchByCityHandler(data: weatherInfo): void {
    dispatch(searchByCityAction(data));
  }
  function searchCityHandler(city: cityInfo): void {
    dispatch(searchCityAction(city));
  }
  function removeCityHandler(city: string): void {
    dispatch(removeCityAction(city));
  }
  function setWeatherDataHandler(data: weatherInfo): void {
    dispatch(setWeatherDataAction(data));
  }
  function errorHandler(error: string): void {
    dispatch(setErrorAction(error));
  }
  function resetSetErrorHandler(): void {
    dispatch(resetErrorAction());
  }

  function refreshState(): void {
    const refreshedState: Array<cityInfo> = localStorage.getItem('state')
      ? JSON.parse(localStorage.getItem('state')!)
      : initialState.city;
    dispatch(refreshStateAction(refreshedState));
  }

  useEffect(() => {
    refreshState();
  }, []);

  useEffect(() => {
    localStorage.setItem('state', JSON.stringify(state.city));
  }, [state]);

  return (
    <GlobalContext.Provider
      value={{
        data: state.data,
        city: state.city,
        error: state.error,
        searchByCityHandler,
        searchCityHandler,
        setWeatherDataHandler,
        errorHandler,
        resetSetErrorHandler,
        removeCityHandler,
      }}
    >
      {children}
    </GlobalContext.Provider>
  );
};

GlobalProvider.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};
