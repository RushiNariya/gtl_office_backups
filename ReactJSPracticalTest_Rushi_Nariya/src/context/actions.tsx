import cityInfo from '../intefaces/cityInfo';
import weatherInfo from '../intefaces/weatherInfo';
import {
  REFRESH_STATE,
  REMOVE_CITY,
  RESET_ERROR,
  SEARCH_BY_CITY,
  SEARCH_CITY,
  SET_ERROR,
  SET_WEATHER_DATA,
} from './actionTypes';
import { initState } from './InitialState';

export interface ActionMethodType {
  type: string;
  payload: weatherInfo | string | initState | Array<cityInfo>;
}

export const searchByCityAction = (data: weatherInfo): ActionMethodType => ({
  type: SEARCH_BY_CITY,
  payload: data,
});

export const setWeatherDataAction = (data: weatherInfo): ActionMethodType => ({
  type: SET_WEATHER_DATA,
  payload: data,
});

export const searchCityAction = (data: cityInfo): ActionMethodType => ({
  type: SEARCH_CITY,
  payload: data,
});

export const removeCityAction = (city: string): ActionMethodType => ({
  type: REMOVE_CITY,
  payload: city,
});

export const setErrorAction = (error: string): ActionMethodType => ({
  type: SET_ERROR,
  payload: error,
});

export const resetErrorAction = (): ActionMethodType => ({
  type: RESET_ERROR,
  payload: '',
});

export const refreshStateAction = (
  state: Array<cityInfo>
): ActionMethodType => ({
  type: REFRESH_STATE,
  payload: state,
});
