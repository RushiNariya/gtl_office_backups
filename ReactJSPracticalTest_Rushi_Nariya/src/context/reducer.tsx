/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import {
  SEARCH_BY_CITY,
  SEARCH_CITY,
  SET_WEATHER_DATA,
  SET_ERROR,
  RESET_ERROR,
  REMOVE_CITY,
  REFRESH_STATE,
} from './actionTypes';
import { initState } from './InitialState';

const AppReducer = (state: initState, action: any): initState => {
  switch (action.type) {
    case SEARCH_BY_CITY:
      return {
        ...state,
      };
    case SET_WEATHER_DATA:
      return {
        ...state,
        data: action.payload,
      };
    case SEARCH_CITY:
      if (state.city.find((item) => item.name === action.payload.name)) {
        return state;
      }
      return {
        ...state,
        city: [...state.city, action.payload],
      };
    case REMOVE_CITY:
      return {
        ...state,
        data: {},
        city: state.city.filter((item) => item.name !== action.payload),
      };
    case SET_ERROR:
      return {
        ...state,
        error: action.payload,
      };
    case RESET_ERROR:
      return {
        ...state,
        error: '',
      };
    case REFRESH_STATE:
      return {
        ...state,
        city: action.payload,
        error: '',
      };
    default:
      return state;
  }
};

export default AppReducer;
