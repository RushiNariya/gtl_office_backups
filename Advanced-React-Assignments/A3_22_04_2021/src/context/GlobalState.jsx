import React, { createContext, useReducer } from 'react';
import PropTypes from 'prop-types';
import reducer, { initialState } from './reducer';
import {
  playAction,
  pauseAction,
  playNextAction,
  playPreviousAction,
  jumpToSlideAction,
} from './ActionMethods';

export const GlobalContext = createContext(initialState);

export const GlobalProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  function playImage() {
    dispatch(playAction());
  }
  function pauseImage() {
    dispatch(pauseAction());
  }
  function playNextImage(id) {
    dispatch(playNextAction(id));
  }
  function playPreviousImage(id) {
    dispatch(playPreviousAction(id));
  }
  function jumpToSlideImage(id) {
    dispatch(jumpToSlideAction(id));
  }

  return (
    <GlobalContext.Provider
      value={{
        state,
        playImage,
        pauseImage,
        playNextImage,
        playPreviousImage,
        jumpToSlideImage,
      }}
    >
      {children}
    </GlobalContext.Provider>
  );
};

GlobalProvider.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};
