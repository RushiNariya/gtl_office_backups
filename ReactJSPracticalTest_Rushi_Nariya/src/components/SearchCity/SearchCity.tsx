import React, { useState, useContext, useEffect } from 'react';
import toast from 'react-hot-toast';
import suggestions from '../../Data/cities';
import { getSearchCityWeatherDataApi } from '../../Api/weatherAPI';
import './SearchCity.css';
import cityInfo from '../../intefaces/cityInfo';
import { GlobalContext } from '../../context/InitialState';

function SearchCity(): JSX.Element {
  const { searchCityHandler, errorHandler, resetSetErrorHandler } =
    useContext(GlobalContext);

  const initData: Array<{ id: number; city: string; }> = [];

  const [searchCity, setSearchCity] = useState('');
  const [showSuggestions, setShowSuggestions] = useState(false);
  const [filteredSuggestions, setFilteredSuggestions] = useState(initData);
  const [activeSuggestion, setActiveSuggestion] = useState(0);
  const [value, setValue] = useState('');

  const handleSubmit = (e): void => {
    e.preventDefault();
    setValue(searchCity);
    setSearchCity('');
  };

  const getWeatherByCity = async (): Promise<any> => {
    const data = await getSearchCityWeatherDataApi(value);

    if (data.cod === 200) {
      resetSetErrorHandler();

      const cityData: cityInfo = {
        name: data.name,
        country: data.sys.country,
        temp: data.main.temp,
        icon: data.weather[0].icon,
        longitude: data.coord.lon,
        latitude: data.coord.lat,
        maxTemp: data.main.temp_max,
        minTemp: data.main.temp_min,
      };
      searchCityHandler(cityData);
    } else {
      toast.error(data.message);
      errorHandler(data.message);
    }
    setSearchCity('');
  };

  useEffect(() => {
    if (value) {
      getWeatherByCity();
    }
  }, [value]);

  const onClick = (e): void => {
    setFilteredSuggestions([]);
    setShowSuggestions(false);
    setActiveSuggestion(0);
    setValue(e.currentTarget.innerText);
    setSearchCity('');
  };

  const onChange = (e): void => {
    setSearchCity(e.target.value);

    if (searchCity.length < 2) {
      setFilteredSuggestions([]);
      setShowSuggestions(false);
      setActiveSuggestion(0);
    }

    if (searchCity.length > 1) {
      const filteredSuggestions = suggestions.filter((suggestion) => {
        if (suggestion.city.toLowerCase().indexOf(searchCity.toLowerCase()) >= 0) {
          return suggestion;
        }
      });
      setFilteredSuggestions(filteredSuggestions);
      setShowSuggestions(true);
      setActiveSuggestion(0);
    }
  };

  const onKeyDown = (e): void => {
    // pressed the enter key
    if (e.keyCode === 13) {
      if (filteredSuggestions.length > 0) {
        setSearchCity(filteredSuggestions[activeSuggestion].city);
      } else {
        setSearchCity(e.target.value);
      }
      setShowSuggestions(false);
      setActiveSuggestion(0);
    }
    // pressed the up arrow
    else if (e.keyCode === 38) {
      if (activeSuggestion === 0) {
        return;
      }
      setActiveSuggestion((preValue) => preValue - 1);
    }
    // pressed the down arrow
    else if (e.keyCode === 40) {
      if (activeSuggestion - 1 === filteredSuggestions.length) {
        return;
      }
      setActiveSuggestion((preValue) => preValue + 1);
    }
  };

  let suggestionsListComponent;

  if (showSuggestions && searchCity) {
    if (filteredSuggestions.length) {
      suggestionsListComponent = (
        <ul className="suggestions">
          {filteredSuggestions.map((suggestion, index) => {
            let className;

            if (index === activeSuggestion) {
              className = 'suggestion-active';
            }
            return (
              <li className={className} key={suggestion.id} onClick={onClick}>
                {suggestion.city}
              </li>
            );
          })}
        </ul>
      );
    } else {
      suggestionsListComponent = (
        <div className="no-suggestions">
          <em>No suggestions, you are on your own!</em>
        </div>
      );
    }
  }

  return (
    <>
      <form onSubmit={handleSubmit}>
        <div className="pseudo-search">
          <input
            type="text"
            required
            placeholder="Search City"
            value={searchCity}
            onChange={onChange}
            onKeyDown={onKeyDown}
          />
          <button className="fa fa-search" type="submit"></button>
        </div>
        {suggestionsListComponent}
      </form>
    </>
  );
}
export default SearchCity;
