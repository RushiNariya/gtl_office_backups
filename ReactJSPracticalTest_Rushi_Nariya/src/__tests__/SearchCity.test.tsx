import React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import SearchCity from '../components/SearchCity/SearchCity';
import { GlobalProvider } from '../context/GlobalState';
import initialState from '../context/InitialState';

let wrapped: ReactWrapper;

beforeEach((): void => {
  wrapped = mount(
    <GlobalProvider initialState={initialState}>
      <SearchCity />
    </GlobalProvider>
  );
});

afterEach((): void => {
  wrapped.unmount();
});

it('renders the SearchCity component', (): void => {
  expect(wrapped).toMatchSnapshot();
});

describe('check input search', (): void => {
  beforeEach((): void => {
    wrapped.find('input').simulate('change', { target: { value: 'New City' } });
    wrapped.update();
  });

  it('check correct data on input click', (): void => {
    expect(wrapped.find('input').prop('value')).toEqual('New City');
  });

  it('check when form is submitted, input search gets emptied', (): void => {
    wrapped.find('form').simulate('submit');
    wrapped.update();
    expect(wrapped.find('input').prop('value')).toEqual('');
  });
});

describe('check form render in SearchCity Component', (): void => {
  it('should render form and button', () => {
    expect(wrapped.find('form').length).toEqual(1);
    expect(wrapped.find('button').length).toEqual(1);
  });
});
