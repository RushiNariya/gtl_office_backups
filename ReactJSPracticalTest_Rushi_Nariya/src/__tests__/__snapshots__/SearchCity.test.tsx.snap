// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the SearchCity component 1`] = `
<GlobalProvider
  initialState={
    Object {
      "city": Array [],
      "data": Object {},
      "error": "",
      "errorHandler": [Function],
      "removeCityHandler": [Function],
      "resetSetErrorHandler": [Function],
      "searchByCityHandler": [Function],
      "searchCityHandler": [Function],
      "setWeatherDataHandler": [Function],
    }
  }
>
  <SearchCity>
    <form
      onSubmit={[Function]}
    >
      <div
        className="pseudo-search"
      >
        <input
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="Search City"
          required={true}
          type="text"
          value=""
        />
        <button
          className="fa fa-search"
          type="submit"
        />
      </div>
    </form>
  </SearchCity>
</GlobalProvider>
`;
